### PR TITLE
Validate state param when talking to External IDPs

### PR DESCRIPTION
**Pull request by andrewedstrom**
As per the spec's requirements for OAuth clients https://tools.ietf.org/html/draft-ietf-oauth-v2-30#section-10.12

Includes a lot of refactors to make the the XOAuthProviderConfigurator
more readable and a complete rewrite of the XOAuthAuthenticationFilter's
tests, as they were growing hard to extend.

[#170636851]

Co-authored-by: Joshua Casey <jcasey@pivotal.io>
Co-authored-by: Andrew Wittrock <awittrock@pivotal.io>
Co-authored-by: Andrew Edstrom <aedstrom@pivotal.io>
Signed-off-by: Andrew Edstrom <aedstrom@pivotal.io>

### CODE REVIEW

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthAuthenticationFilter.java [hunk 1]
```diff
@@ -89,6 +109,9 @@ private boolean authenticationWasSuccessful(
             ofNullable(successHandler).ifPresent(handler ->
                     handler.setSavedAccountOptionCookie(request, response, authentication)
             );
+        // TODO: :eyes_narrowed:
```
Comment (andrewedstrom) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthAuthenticationFilter.java:112]: We definitely need to remove this TODO before merging

Comment (shamus) [reply server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthAuthenticationFilter.java:112]: But why not fix it? Catching all exceptions scary.

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthProviderConfigurator.java [hunk 2]
```diff
@@ -48,31 +52,56 @@ protected OIDCIdentityProviderDefinition overlay(OIDCIdentityProviderDefinition
...
+            uriBuilder.queryParam("nonce", nonceGenerator.generate());
+        }
+
+        return uriBuilder.build().toUriString();
+    }
+
+    private void saveStateParamForIdpToRequestSession(String state, String idpOriginKey, HttpServletRequest request) {
+        if (request != null) {
+            request.getSession().setAttribute("xoauth-state-" + idpOriginKey, state);
+        }
```
Comment (shamus) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthProviderConfigurator.java]: Is there any value to this null check? Even if it is null, this operation doesn't make any sense if we can't access the session.

Comment (andrewedstrom) [reply server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthProviderConfigurator.java]: Good point

**Pull request merged**